tpm2/structures.go

```diff
@@ -1028,3 +1029,70 @@ type TPMLDigest struct {
 	Count   uint32
 	Digests []tpmutil.U16Bytes
 }
+
+// Encode converts the TPMLDigest structure into a byte slice
+func (list *TPMLDigest) Encode() ([]byte, error) {
+	if uint32(len(list.Digests)) != list.Count {
+		return nil, fmt.Errorf("Count invalid. Have: %v - Want: %v", len(list.Digests), list.Count)
+	}
+	res, err := tpmutil.Pack(list.Count)
+	if err != nil {
+		return nil, err
+	}
+	for _, item := range list.Digests {
+		b, err := tpmutil.Pack(item)
+		if err != nil {
+			return nil, err
+		}
+		for _, i := range b {
+			res = append(res, i)
+		}
+	}
+	return res, nil
+}
+
+// DecodeTPMLDigest decodes a TPML_Digest part of a message. No error is returned if
+// the input has extra trailing data.
+func DecodeTPMLDigest(buf []byte) (*TPMLDigest, error) {
+	in := bytes.NewBuffer(buf)
+	var tpmld TPMLDigest
+	var err error
+	var hashlen uint16
+	if err = binary.Read(in, binary.BigEndian, &tpmld.Count); err != nil {
+		return &tpmld, fmt.Errorf("decoding TPML_Digest: %v", err)
+	}
+
+	for i := uint32(0); i < tpmld.Count; i++ {
```

Comment:
I think this code could be simplified to take advantage of U16Bytes.TPMUnmarshal, which handles the mechanics of reading the size and byte array:
https://github.com/google/go-tpm/blob/240a3433a7a53049ab3386b7dace1e4d90dba2ce/tpmutil/structures.go#L56-L78

Here, I'd propose only checking the `Count` that you read at the beginning against the total number of `U16Bytes`es that you deserialized from the slice, something like:

```
for in.Len() > 0 {
  var d tpmutil.U16Bytes
  if err = d.TPMUnmarshal(in); err != nil {
    return nil, err
  }
  tpmld.Digests = append(tpmld.Digsts, d)
}
if Count != len(tpmld.Digests) {
  return nil, fmt.Errorf("blah blah blah")
}